aten/src/ATen/native/ComparisonUtils.cpp

```diff
@@ -0,0 +1,29 @@
+#include <ATen/core/TensorBase.h>
+#include <algorithm>
+#include <vector>
+#include <ATen/core/TensorBody.h>
+
+namespace at {
+
+class Tensor;
+
+namespace native {
+
+void _assert_tensor_metadata(at::Tensor const& tensor, c10::OptionalArrayRef<long> sizes, c10::OptionalArrayRef<long> strides, c10::optional<c10::ScalarType> dtype) {
+  std::stringstream msg;
+  if (sizes) {
+    msg << "Tensor size mismatch! Got " << tensor.sizes().vec() << " expected" << sizes.value().vec() << std::endl;
```

Comment:
Is constructing these strings expensive?  Should we check the condition before constructing them?

Reply:
I assumed it would get inlined and optimized into the condition check by the compiler, but it would not hurt to be explicit.

Reply:
I doubt it will because string construction calls `new`, though you could check the output assembly.